aio-interface: Improve psalm

docjyJ opened the pull request:
### Fix warnings generated by phpstorm

- Add `psalm:strict` until type and unused class warnings are fixed (a lot `277 other issues found.`)
- https://psalm.dev/docs/running_psalm/configuration/
- ![image](https://github.com/user-attachments/assets/a1f65a2d-e010-448d-a85d-0531935ed20a)

docjyJ commented:
```
Found 43 of 43 files that can be fixed in 6.985 seconds, 30.00 MB memory used
Please run `composer run cs:fix` to format your code
Error: Process completed with exit code 1.
```
Perfect, the workflow works, it fails successfully ^^'

PR ready

szaimen commented:
> Perfect, the workflow works, it fails successfully ^^'

I fear we cannot merge this with red ci. Can you run cs fix and commit the changes in a separate commit?

Also adding psalm:sctrict seems to be unrelated? Can you out this into a separate pr please?

docjyJ commented:
> I fear we cannot merge this with red ci. Can you run cs fix and commit the changes in a separate pr that builds on this branch?

I know...

> Also the changes to psalm seem to be unrelated? Can you put this into a separate pr please?

I don't know, it is link with `<!--TODO: errorLevel="1" see PR #5369-->`.

szaimen commented on php/composer.json:
```diff
@@ -1,37 +1,42 @@
+		],
 		"psalm": "psalm --threads=1",
 		"psalm:update-baseline": "psalm --threads=1 --monochrome --no-progress --output-format=text --update-baseline",
+		"psalm:strict": "psalm --threads=1 --show-info=true",
```
lets put this into a different PR please

szaimen commented on php/psalm.xml:
```diff
@@ -1,17 +1,23 @@
-	errorLevel="2"
-	resolveFromConfigFile="true"
 	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
 	xmlns="https://getpsalm.org/schema/config"
-	xsi:schemaLocation="https://getpsalm.org/schema/config"
-    errorBaseline="psalm-baseline.xml"
+	xsi:schemaLocation="https://getpsalm.org/schema/config vendor/vimeo/psalm/config.xsd"
+	errorBaseline="psalm-baseline.xml"
 	findUnusedBaselineEntry="true"
-	findUnusedCode="false"
 >
 	<projectFiles>
 		<directory name="templates"/>
 		<directory name="src"/>
 		<file name="public/index.php"/>
+		<ignoreFiles>
+			<directory name="vendor" />
+		</ignoreFiles>
 	</projectFiles>
+	<extraFiles>
+		<directory name="vendor" />
+	</extraFiles>
+	<issueHandlers>
+	</issueHandlers>
```
the changes here as well please

docjyJ commented:
ok psalm done

docjyJ commented:
ok see #5370 for CS

szaimen commented on php/composer.json:
```diff
@@ -1,35 +1,36 @@
+		],
 		"psalm": "psalm --threads=1",
 		"psalm:update-baseline": "psalm --threads=1 --monochrome --no-progress --output-format=text --update-baseline",
+		"psalm:strict": "psalm --threads=1 --show-info=true",
```
should we also update the readme to describe this new option?

docjyJ commented:
```
------------------------------
284 errors found
------------------------------
Psalm can automatically fix 3 of these issues.
Run Psalm again with 
--alter --issues=UnusedVariable --dry-run
to see what it can fix.
------------------------------
```
In commit c20a8f8
I switched psalm to error level 1. The strict command is no longer useful, so I removed it.

The problem is that the workflow fails.

This is due to some problems in the code.

Ideally it would be good to switch to level 1. This avoids using variables of type `mixed` and unused variables. IDEs can be more efficient for autocompletion, and psalm can warn of type errors more easily.

docjyJ commented:
> Seems like static analysis is failing now :/
> 
> Can we increase the error level in a follow-up?

That's why I used a TODO comment and added the stric composer script.
The idea is to improve the code little by little and at some point definitely move to level 1.

docjyJ commented:
Issue in code:
- [MixedArgument](https://psalm.dev/docs/running_psalm/issues/MixedArgument/)
- [MixedArrayAccess](https://psalm.dev/docs/running_psalm/issues/MixedArrayAccess/)
- [MixedArrayAssignment](https://psalm.dev/docs/running_psalm/issues/MixedArrayAssignment/)
- [MixedArrayOffset](https://psalm.dev/docs/running_psalm/issues/MixedArrayOffset/)
- [MixedAssignment](https://psalm.dev/docs/running_psalm/issues/MixedAssignment/)
- [MixedInferredReturnType](https://psalm.dev/docs/running_psalm/issues/MixedInferredReturnType/)
- [MixedMethodCall](https://psalm.dev/docs/running_psalm/issues/MixedMethodCall/)
- [MixedOperand](https://psalm.dev/docs/running_psalm/issues/MixedOperand/)
- [MixedPropertyTypeCoercion](https://psalm.dev/docs/running_psalm/issues/MixedPropertyTypeCoercion/)
- [MixedReturnStatement](https://psalm.dev/docs/running_psalm/issues/MixedReturnStatement/)
- [PossiblyNullOperand](https://psalm.dev/docs/running_psalm/issues/PossiblyNullOperand/)
- [UnusedClosureParam](https://psalm.dev/docs/running_psalm/issues/UnusedClosureParam/)
- [UnusedVariable](https://psalm.dev/docs/running_psalm/issues/UnusedVariable/)

Command used:
```bash
composer run psalm -- --monochrome --no-progress | grep ^ERROR | sed -e 's/^ERROR: //' | sort -u
```

szaimen commented:
> > Seems like static analysis is failing now :/
> > Can we increase the error level in a follow-up?
> 
> That's why I used a TODO comment and added the stric composer script. The idea is to improve the code little by little and at some point definitely move to level 1.

Ah I see. Can we for now switch back to errorlevel2 and then in follow-ups resolve the issues?

docjyJ commented:
I drop the commit

docjyJ commented:
I don't know if you've seen it yet, but I've started working on stalwart integration. I'm trying to apply strict rules in static analysis.

https://github.com/docjyJ/stalwart-connector

_(I recently got into rust and really enjoyed it. I'm a bit of a neat freak, but I think it's important to avoid errors. :sweat_smile:)_

docjyJ commented:
Follow up #5368

docjyJ commented:
I found that in psalm doc: https://psalm.dev/docs/manipulating_code/fixing/

szaimen reviewed (approved):
LGTM, thanks 😊

szaimen commented:
> I found that in psalm doc: https://psalm.dev/docs/manipulating_code/fixing/

This sounds cool as an additonal commans maybe? :)

szaimen commented:
> I don't know if you've seen it yet, but I've started working on stalwart integration. I'm trying to apply strict rules in static analysis.
> 
> https://github.com/docjyJ/stalwart-connector
> 
> _(I recently got into rust and really enjoyed it. I'm a bit of a neat freak, but I think it's important to avoid errors. 😅)_

Also this looks very interesting! Cool that you started working on it 😊

docjyJ commented:
> > I found that in psalm doc: https://psalm.dev/docs/manipulating_code/fixing/
> 
> This sounds cool as an additonal command maybe? :)

I think this should be added globally to the Nextcloud style rule to maintain consistency.

docjyJ commented:
> > I don't know if you've seen it yet, but I've started working on stalwart integration. I'm trying to apply strict rules in static analysis.
> > https://github.com/docjyJ/stalwart-connector
> > _(I recently got into rust and really enjoyed it. I'm a bit of a neat freak, but I think it's important to avoid errors. 😅)_
> 
> Also this looks very interesting! Cool that you started working on it 😊

Thank you! If you have time, don't hesitate to give feedback.

szaimen commented:
This is now released with v9.7.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel
